applications/blogs/backend/src/main/java/pl/ds/kyanite/blogs/components/models/BlogPostTableOfContentsComponent.java

```diff
@@ -74,27 +86,71 @@ public class BlogPostTableOfContentsComponent {
 
   @PostConstruct
   private void init() {
-    this.items = new ArrayList<>();
+    prepareTitlesHierarchy();
+  }
 
+  private void prepareTitlesHierarchy() {
+    this.subTitles = new ArrayList<>();
     Resource pageResource = getContainingPageAsResource(this.resource);
 
     if (pageResource != null) {
       Resource contentsContainer = resourceResolver.getResource(
           pageResource.getPath() + BLOG_CONTENT_CONTAINER_REL_PATH);
-      List<Resource> titles = findAllResourcesInContainerByType(
-          contentsContainer, TITLE_RESOURCE_PATH);
-
-      for (Resource contentTitle : titles) {
-        TitleComponent titleComponent = contentTitle.adaptTo(TitleComponent.class);
-
-        if (!"p".equals(titleComponent.getElement())) {
-          BlogPostTableOfContentsItemComponent contentItem =
-              prepareNewContentItem(titleComponent);
-          if (contentItem != null) {
-            this.items.add(contentItem);
+      List<Resource> titles = findAllValidTitlesInContainer(contentsContainer);
+
+      hierarchyErrorMessage = "";
+      ListIterator<Resource> titlesIter = titles.listIterator();
+      BlogPostTableOfContentsItemComponent stubRoot = new BlogPostTableOfContentsItemComponent(
+          "stub", "", "");
+      BlogPostTableOfContentsItemComponent currentNode = stubRoot;
+      int currentLevel = maxHeadingLevel - 1;
+
+      while (titlesIter.hasNext()) {
```

Comment:
This `prepareTitlesHierarchy` is pretty long. Maybe if you will be able to apply changes from above comments then this while could be simplified to something like
```
while (titlesIter.hasNext()) {
        Resource titleResource = titlesIter.next();
        currentNode = handleAddTitleNode(titleResource);
}        
```
?